### PR TITLE
Sync olp::utils::LruCache with hf::container::LruCache

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Make implementation of olp::utils::LruCache similar
to hf::container::LruCache. Change usage of custom
hf::container::HashFunction to std::hash function.

Relates-To: OLPEDGE-1590

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/include/olp/core/utils/EmptyBaseClassOptimizer.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  explicit EmptyBaseClassOptimizer(T data) : m_data(data) {}
+
+  T& data() { return m_data; }
+
+  const T& data() const { return m_data; }
+
+ private:
+  T m_data;
+};
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/EmptyBaseClassOptimizer.h:68]: Please adapt to coding style.

File: olp-cpp-sdk-core/include/olp/core/utils/LruCache.h [hunk 2]
```diff
@@ -19,592 +19,479 @@
...
+#include <assert.h>
 #include <cstddef>
 #include <functional>
 #include <map>
 
 #include <olp/core/porting/try_emplace.h>
-
+#include "NullEvictionCallback.h"
+#include "UnitWeightFunction.h"
+#include "UnorderedMap.h"
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/LruCache.h:30]: Use <> to include in public headers.

File: olp-cpp-sdk-core/include/olp/core/utils/LruCache.h [hunk 3]
```diff
@@ -19,592 +19,479 @@
...
+          class EvictionCallback = NullEvictionCallback,
+          typename Index = uint32_t,
+          template <typename, typename> class Cont = std::vector,
           template <typename> class Alloc = std::allocator>
 class LruCache {
-  struct Bucket;
-  using MapType =
-      std::map<Key, Bucket, Compare, Alloc<std::pair<const Key, Bucket> > >;
+ private:
+  static constexpr Index HEAD_INDEX = 0;
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/LruCache.h:66]:
```suggestion
  static constexpr Index kHeadIndex = 0;
```

File: olp-cpp-sdk-core/include/olp/core/utils/LruCache.h [hunk 4]
```diff
@@ -19,592 +19,479 @@
...
+ *                         The function signature is expected to be void( const
+ * Key&, Value&& ).
+ * @param Index Index type used internally in the underlying hash map.
+ *              Make sure that it supports the amount of elements you want to
+ * insert
  */
-template <typename Key, typename Value,
-          typename CacheCostFunc = CacheCost<Value>,
-          typename Compare = std::less<Key>,
+template <typename Key, typename Value, class WeightFunc = UnitWeightFunction,
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/LruCache.h:58]: Please adapt to coding style.

File: olp-cpp-sdk-core/include/olp/core/utils/LruCache.h [hunk 5]
```diff
@@ -19,592 +19,479 @@
...
+                    EvictionCallback eviction_func = EvictionCallback(),
+                    const AllocType& alloc = AllocType())
+      : m_items_map(0, alloc),
+        m_head(),
+        m_max_weight(max_weight),
+        m_weight(0),
+        m_eviction_func(std::move(eviction_func)),
+        m_weight_func(std::move(weight_function)) {}
+
+  ~LruCache() {
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/LruCache.h:173]:
```suggestion
  ~LruCache()  = default;
```

File: olp-cpp-sdk-core/include/olp/core/utils/LruCache.h [hunk 6]
```diff
@@ -19,592 +19,479 @@
...
+    bool inserted;
+    std::tie(iter, inserted) = m_items_map.try_emplace(
+        std::forward<KeyArg>(key), std::forward<T>(value));
+    Index index = GetIndex(iter);
+    if (!inserted) {
+      ListUnlink(iter);
+      ListPushFront(iter);
+      return std::make_pair(const_iterator(index, *this), false);
+    }
+    assert(m_weight_func(iter->data().m_value) == weight &&
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/LruCache.h:232]: I would prefer we do not use asserts. Use if loop if it is a likely error.

**Comment by LiubovDidkivska** [comment]
> This is a huge API break. Can we not just sync internals, instead of everything?

Updated. API is adjusted to current one.

File: olp-cpp-sdk-core/include/olp/core/utils/LruCache.h [hunk 7]
```diff
@@ -40,217 +39,303 @@ template <typename T>
...
+ * doubly-linked list, exploiting the fact that the order of the entries in the
+ * UnorderedMap is stable under most operations (excluding erase).
+ *
+ * @param Key Type of the key used to access a cached element.
+ * @param Value Type of the value stored for a cached element.
+ * @param CacheCostFunc Functor that maps a value to its cost.
+ * The cost of a value must never change while the value is stored in the cache.
+ *
+ * @param HashFunc Functor that maps a key to a hash value.
+ * @param Index Index type used internally in the underlying hash map.
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/LruCache.h:64]: Template parameters should be documented with `@tparam`

**Comment by LiubovDidkivska** [comment]
> This commit is too huge. Please split up.
> Also please consider using std::unordered_map instead of the internal UnorderedMap. I dont think we need the full speed of the internal UnorderedMap which is mostly designed for tuff performance uses case (Routing or rendering) which we will not face for now.

Changing to std::unordered_map will affect implementation of LruCache, as UnorderedMap implemented some advanced futures for example indexing, and have some differences in API compared to std::unordered_map. We can implement using std::unordered_map, but it could look more similar to current LruCache implementation.

**Pull request closed without merging**